File: GameEngine/src/Engine/Math/Random.cpp
#include "enginepch.h"
#include "Engine/Math/Random.h"

namespace Engine
{
	namespace Math
	{
		static std::mt19937 RandomEngine(unsigned int seed)
		{
			return std::mt19937(seed);
		}

		float Random::Float()
		{
			return Float(std::random_device()());
		}

		float Random::Float(unsigned int seed)
		{
			std::uniform_real_distribution<float> distribution;
			return distribution(RandomEngine(seed));
		}

		int Random::Int()
		{
			return Int(std::random_device()());
		}

		int Random::Int(unsigned int seed)
		{
			std::uniform_int_distribution<std::mt19937::result_type> distribution;
			return distribution(RandomEngine(seed));
		}

		float Random::Range(float min, float max)
		{
			return Range(min, max, std::random_device()());
		}


		float Random::Range(float min, float max, unsigned int seed)
		{
			std::uniform_real_distribution<float> distribution(min, max);
			return distribution(RandomEngine(seed));
		}


		int Random::Range(int min, int max)
		{
			return Range(min, max, std::random_device()());
		}


		int Random::Range(int min, int max, unsigned int seed)
		{
			std::uniform_int_distribution<std::mt19937::result_type> distribution(min, max);
			return distribution(RandomEngine(seed));
		}
	}
}

File: Engine-ScriptCore/Source/Engine/Math/Random.cs
using Engine.Core;

namespace Engine.Math
{
	public class Random
	{
		// 0 - 1, max exclusive, uniform distribution
		public static float Float() => InternalCalls.Random_Float();
		public static float Float(uint seed) => InternalCalls.Random_Float_Seed(seed);

		// any int, max inclusive, uniform distribution
		public static int Int() => InternalCalls.Random_Int();
		public static int Int(uint seed) => InternalCalls.Random_Int_Seed(seed);

		// max exclusive, uniform distribution
		public static float Range(float min, float max) => InternalCalls.Random_Range_Float(min, max);
		public static float Range(float min, float max, uint seed) => InternalCalls.Random_Range_Float_Seed(min, max, seed);

		// max inclusive, uniform distribution
		public static int Range(int min, int max) => InternalCalls.Random_Range_Int(min, max);
		public static int Range(int min, int max, uint seed) => InternalCalls.Random_Range_Int_Seed(min, max, seed);
	}
}
